File: src/main/java/de/thelooter/eventchecker/commands/complete/EventCheckerCommandCompleter.java
package de.thelooter.eventchecker.commands.complete;

import com.google.common.collect.Lists;
import de.thelooter.eventchecker.EventChecker;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class EventCheckerCommandCompleter implements TabCompleter {
    @Override
    public List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label,
                                      @NotNull String[] args) {
        if (args.length == 1) {
            return List.of("reload", "list");
        }

        if (args.length == 2) {
            if (args[0].equals("list")) {
                return List.of("all");
            }
        }

        if (args.length == 3) {
            if (args[0].equals("list")) {
                if (args[1].equals("all")) {
                    List<Integer> pages = new ArrayList<>();
                    for (int i = 0; i < Lists.partition(EventChecker.eventNames, 50).size(); i++) {
                        pages.add(i+1);
                    }
                    return pages.stream().map(String::valueOf).toList();
                }

            }
        }

        return Collections.emptyList();
    }
}

File: src/main/kotlin/de/thelooter/eventchecker/commands/complete/EventCheckerCommandCompleter.kt
package de.thelooter.eventchecker.commands.complete

import com.google.common.collect.Lists
import de.thelooter.eventchecker.EventChecker
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter

/**
 * The TabCompleter for the EventChecker Command
 * @see TabCompleter
 * @since 1.2.0
 */
class EventCheckerCommandCompleter : TabCompleter {
    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<out String>?
    ): List<String> {

        args?.let {

            if (it.size == 1) {
                return listOf("reload", "list")
            }

            if (it.size == 2 && it[0] == "list") {
                return listOf("all", "blacklist", "whitelist")
            }

            if (it.size == 3 && it[0] == "list") {
                val pageSize = 25
                return when (it[1]) {
                    "all" -> createPageList(EventChecker.eventNames, pageSize)
                    "blacklist" -> {
                        createPageList(
                            EventChecker.instance.config.getStringList("excluded-events"),
                            pageSize
                        )
                    }

                    "whitelist" -> {
                        createPageList(
                            EventChecker.instance.config.getStringList("included-events"),
                            pageSize
                        )
                    }

                    else -> emptyList()
                }
            }

        }
        return emptyList()
    }


    /**
     * Creates a list of page numbers based on a given list and page size.
     *
     * @param list The input list of elements.
     * @param pageSize The number of elements per page.
     * @return The list of page numbers as strings.
     */
    private fun createPageList(list: List<String>, pageSize: Int): List<String> {
        return Lists.partition(list, pageSize).indices.map { (it + 1).toString() }
    }
}